lib/wibox/layout/align.lua

```diff
@@ -35,127 +35,336 @@ local base = require("wibox.widget.base")
-            --  the remaining space if available we will draw later
-            size_remains = floor((size_remains - size_second) / 2)
+    local first = self._private.first
+    local second = self._private.second
+    local third = self._private.third
+
+    -- Wrapper around layout_fit_widget function to pass less parameters
+    local function fit_widget(widget, max_size)
+        -- Can't place a widget if it does not exist
+        if not widget then
+            return 0
         end
+        -- Default values
+        max_size = max_size and max_size or size_remains
+
+        return layout_fit_widget(
+            self, context, width, height,
+            widget, max_size
+        )
     end
-    if self._private.first then
-        local w, h, _ = width, height, nil
-        -- we use the fit function for the "inside" and "none" modes, but
-        --  ignore it for the "outside" mode, which will force it to expand
-        --  into the remaining space
-        if self._private.expand ~= "outside" then
-            if self._private.dir == "y" then
-                _, h = base.fit_widget(self, context, self._private.first, width, size_remains)
-                size_first = h
-                -- for "inside", the third widget will get a chance to use the
-                --  remaining space, then the middle widget. For "none" we give
-                --  the third widget the remaining space if there was no second
-                --  widget to take up any space (as the first if block is skipped
-                --  if this is the case)
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - h
-                end
-            else
-                w, _ = base.fit_widget(self, context, self._private.first, size_remains, height)
-                size_first = w
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - w
-                end
-            end
+
+    -- Wrapper around layout_place_widget function to pass less parameters
+    local function place_widget(widget, offset, size)
+        -- Can't place a widget if it does not exist or there is no more size
+        if not widget or size_remains <= 0 then
+            return
+        end
+        -- Default values
+        if not size then
+            size = fit_widget(widget, size_remains)
         else
-            if self._private.dir == "y" then
-                h = size_remains
-            else
-                w = size_remains
-            end
+            fit_widget(widget, size)
         end
-        table.insert(result, base.place_widget_at(self._private.first, 0, 0, w, h))
+
+        local placed = layout_place_widget(
+            self, width, height,
+            widget, offset, size
+        )
+        table.insert(result, placed)
+        size_remains = size_remains - size
```

Comment:
I think this function should return `placed` (and `size_remains`?). Then, it would be the responsibility of the caller to do the insert.

I can understand we want to keep some level of closure to wrap the function call, but I just don't like a closure to have side effect and alter the result directly.

Reply:
Makes sense, I will change the return of the function and the code where I call it accordingly.